python/ciqueue/pytest.py

```diff
@@ -80,13 +92,26 @@ def pytest_runtest_makereport(self, item, call):
             else:
                 item.error_reports[call.when] = payload
 
-            self.redis.hset(
-                self.errors_key,
-                test_queue.key_item(item),
-                dill.dumps(item.error_reports))
-        # if the test passed, we remove it from the errors queue
-        elif call.when == 'teardown' and not hasattr(item, 'error_reports'):
-            self.redis.hdel(self.errors_key, test_queue.key_item(item))
+        if call.when == 'teardown':
+            test_name = test_queue.key_item(item)
+            test_failed = outcomes.failed(item)
+
+            # Only attempt to requeue if the test failed.
+            # The method will return `False` if the test couldn't be requeued
+            if test_failed and self.queue.requeue(test_name):
+                outcomes.mark_as_skipped(call, item, self.terminalreporter.stats, "RETRYING")
+                self.terminalwriter.write(' RETRYING ', green=True)
+
+            # If the test was already acknowledged by another worker (we timed out)
+            # Then we only record it if it was successful.
+            elif self.queue.acknowledge(test_name) or not test_failed:
+                self.record(item)
+
+            # The test timed out and failed, mark it as skipped so that it doesn't
+            # fail the build
+            else:
+                outcomes.mark_as_skipped(call, item, self.terminalreporter.stats, "TIMED OUT")
+                self.terminalwriter.write(' TIMED OUT ', green=True)
```

Comment:
Why are we marking a test that isn't acknowledged (does this mean not running somewhere?) and failed as skipped?

Reply:
So, if `acknowledge()` returned false, it mean another worker already acknowledged that test.

So based on the previous condition: `if self.queue.acknowledge(test_name) or not test_failed`

 - If the test was already acknowledged but is a success -> we record it because it either change nothing or flip a failure into a success.
  - If the test was already acknowledged but is a failure -> we drop it because it either change nothing or flip a success into a failure.

I know that if / else is a bit hard to grasp, that's why it's commented even in the ruby version. There might be a more readable way to express it but I haven't figured it out.

Reply:
So the condition where a test has failed for its 3rd time (i.e. `requeue` will return `False`) should return `True` on `acknowledge()` if no other worker has acknowledged it and it will get recorded on L108 as having failed, correct?

Reply:
> a test has failed for its 3rd time

Assuming the requeue limit per test is configure at 2



> if no other worker has acknowledged it and it will get recorded on L108 as having failed, correct?

Correct.

Reply:
Thanks for the explanation 😄  Yeah that was a bit to unpack. I'm not sure how to make that clearer other than more comments.